pkg/controller/seed-controller-manager/constraint-controller/controller.go

```diff
@@ -0,0 +1,276 @@
+		workerNameLabelSelector: workerSelector,
+		namespace:               namespace,
+		seedClient:              mgr.GetClient(),
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Cluster{}},
+		enqueueConstraints(reconciler.seedClient, reconciler.log, namespace),
+		workerlabel.Predicates(workerName),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for clusters: %w", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Constraint{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.ByNamespace(namespace),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for constraints: %v", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the kubermatic constraints in the seed cluster and syncs them to all user clusters namespace
+// which have opa integration enabled
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Reconciling")
+
+	err := r.reconcile(ctx, log, request)
+	if controllerutil.IsCacheNotStarted(err) {
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+	}
+
+	return reconcile.Result{}, err
+}
+
+func constraintCreatorGetter(constraint *kubermaticv1.Constraint) reconciling.NamedKubermaticV1ConstraintCreatorGetter {
+	return func() (string, reconciling.KubermaticV1ConstraintCreator) {
+		return constraint.Name, func(c *kubermaticv1.Constraint) (*kubermaticv1.Constraint, error) {
+			c.Name = constraint.Name
+			c.Spec = constraint.Spec
+			return c, nil
+		}
+	}
+}
+
+func (r *reconciler) patchFinalizer(ctx context.Context, constraint *kubermaticv1.Constraint, action string) error {
+	oldconstraint := constraint.DeepCopy()
+
+	if action == AddAction {
+		kuberneteshelper.AddFinalizer(constraint, finalizer)
+	} else if action == RemoveAction {
+		kuberneteshelper.RemoveFinalizer(constraint, finalizer)
+	}
+
+	if err := r.seedClient.Patch(ctx, constraint, ctrlruntimeclient.MergeFrom(oldconstraint)); err != nil {
+		return fmt.Errorf("failed to %s constraint finalizer %s: %v", action, constraint.Name, err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+
+	constraint := &kubermaticv1.Constraint{}
+
+	if err := r.seedClient.Get(ctx, request.NamespacedName, constraint); err != nil {
+		return ctrlruntimeclient.IgnoreNotFound(err)
+	}
+
+	if !constraint.DeletionTimestamp.IsZero() {
+		if !kuberneteshelper.HasFinalizer(constraint, finalizer) {
+			return nil
+		}
+
+		if err := r.cleanupConstraint(ctx, log, constraint); err != nil {
+			return err
+		}
```

Comment:
you need to add `return nil` here so the reconcile finishes here if its being deleted

Reply:
@lsviben  You mean something like that?
```
		if err := r.cleanupConstraint(ctx, log, constraint); err != nil {
                          if kerrors.IsNotFound(err) {
                            return nil
                          } 
			return err
		}
```
In such case, I do agree.

Reply:
I meant the `return nil` below which is already added. 

But your suggestion is also good. Although I think NotFound is already handled in cleanupConstraints func

Reply:
ok, thx for the clarification